content/6_Advanced/Count_Min.mdx

```diff
@@ -3,38 +3,206 @@ id: count-min
 
 <Spoiler title="Solution">
 
-<IncompleteSection>
+We can use techniques introduced in [Range Queries with Sweep Line](https://usaco.guide/plat/range-sweep?lang=cpp)
+
+We sweep from left to right over the $x$-coordinates, maintaining two events for each rectangle: one for the left boundary and one for the right boundary. Maintain a Lazy Segment Tree over the $y$-coordinates. 
+- When we run into a left boundary of some rectangle with $y$-coordinates $(y_0, y_1)$, increase each index $i \in [y_0, y_1]$ by 1
+- When we run into a right boundary of some rectangle with $y$-coordinates $(y_0, y_1)$, decrease each index $i \in [y_0, y_1]$ by 1
+
+Then, for each $x$, we simply need to count the number of non-zero indices which corresponds to indices that are covered by at least one rectangle. How can we do this?
 
-describe
+Instead of counting the area covered by at least one rectangle, let's count the amount of space covered by no rectangles. We can subtract this amount from the total number of indices to get the value we want. 
 
-</IncompleteSection>
+We can use a Segment Tree that counts the number of occurences of the minimum value. Because the minimum value is at least zero (there can't be a negative number of rectangles at a position) the number of uncovered squares is equal to the number of squares with value 0.
 
 </Spoiler>
 
+### Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+ 
+struct event {
+	int t, x, y0, y1;
+ 
+	bool operator<(const event& e) { return x < e.x; }
+};
+ 
+int N;
+vector<event> E;
+array<pair<long long, long long>, 8000005> T;
+array<int, 8000005> L;
+ 
+pair<long long, long long> merge(pair<long long, long long> a, pair<long long, long long> b) {
+	if (a.first < b.first)
+		return a;
+	if (a.first > b.first)
+		return b;
+	return MP(a.first, a.second + b.second);
+}
+ 
+void pushdown(int t) {
+	if (L[t]) {
+		T[t << 1].first += L[t];
+		L[t << 1] += L[t];
+		T[t << 1 | 1].first += L[t];
+		L[t << 1 | 1] += L[t];
+		L[t] = 0;
+	}
+}
+ 
+void build(int t = 1, int tl = 0, int tr = 2000000) {
+	if (tl == tr) {
+		T[t] = MP(0, 1);
+		return;
+	}
+	int tm = (tl + tr) >> 1;
+	build(t << 1, tl, tm);
+	build(t << 1 | 1, tm + 1, tr);
+	T[t] = merge(T[t << 1], T[t << 1 | 1]);
+}
+ 
+void update(int l, int r, int v, int t = 1, int tl = 0, int tr = 2000000) {
+	if (r < tl || tr < l)
+		return;
+	if (l <= tl && tr <= r) {
+		T[t].first += v;
+		L[t] += v;
+		return;
+	}
+	if (tl != tr)
+		pushdown(t);
+	int tm = (tl + tr) >> 1;
+	update(l, r, v, t << 1, tl, tm);
+	update(l, r, v, t << 1 | 1, tm + 1, tr);
+	T[t] = merge(T[t << 1], T[t << 1 | 1]);
+}
+ 
+long long query() {
+	return 2000001 - T[1].second;
+} 
+ 
+int main() {
+	ios_base::sync_with_stdio(0); cin.tie(0);
+ 
+	cin >> N; 
+	build();
+	E.reserve(2 * N);
+	for (int i = 0; i < N; i++) {
+		int x0, y0, x1, y1;
+		cin >> x0 >> y0 >> x1 >> y1;
+		x0 += 1000000, y0 += 1000000;
+		x1 += 1000000, y1 += 1000000;
```

Comment:
uh what's this for

Reply:
Handle negative coordiantes

Reply:
100% add a comment for that
also put it in a constant or smth

Reply:
Is this better?